refactor(fluid-build): detect enclosing workspace or package as root if config not found

curtisman opened the pull request:
These are changes discovered while trying to use `fluid-build` in a non-fluid repo, but good general improvements.

Reduce requirement to use `fluid-build` in other repo by not requiring a config (`fluidBuild.config.[cm]?js` or fluidBuild property in `package.json`) to use `fluid-build`.   Instead detect enclosing package/workspace and use it as the root.

curtisman committed:
fluid-build: detect enclosing workspace or package as root if config not found

curtisman committed:
fix merge error

curtisman committed:
Merge branch 'main' into rootr

tylerbutler commented on build-tools/packages/build-tools/src/common/fluidUtils.ts:
```diff
@@ -31,35 +32,72 @@ async function isFluidRootPackage(dir: string) {
+			if (!buildRoot) {
+				return gitRoot;
+			}
+			// For build root, we require it to have fluidBuild property.
```
Why is this?

curtisman replied:
Just didn't want to change behavior for non-fluid build scenarios. But probably can remove it if you think it is fine to change it.

tylerbutler commented on build-tools/packages/build-tools/src/common/fluidUtils.ts:
```diff
@@ -31,35 +32,72 @@ async function isFluidRootPackage(dir: string) {
+			const { rootDir } = await getPackages(process.cwd());
+			return rootDir;
+		} catch (e) {
+			traceInit(`Error getting packages: ${e}`);
```
Maybe "error getting workspace?" or "error enumerating workspace packages?"

curtisman replied:
Changed. Although, note that this works on single packages as well.

tylerbutler commented on build-tools/packages/build-tools/src/common/fluidUtils.ts:
```diff
@@ -31,35 +32,72 @@ async function isFluidRootPackage(dir: string) {
 }
 
-export async function getResolvedFluidRoot() {
+async function inferFluidRoot(buildRoot: boolean) {
```
What's the difference between inferRoot and inferFluidRoot? Why are both needed?

curtisman replied:
I refactor the code into two functions so that I can do early return pattern in `inferFluidRoot`, so that we can add a final check for all the possible returns from `inferRoot`

tylerbutler commented on build-tools/packages/build-tools/src/common/fluidUtils.ts:
```diff
@@ -98,19 +136,32 @@ const configExplorer = cosmiconfigSync("fluidBuild", {
  * @param noCache - If true, the config cache will be cleared and the config will be reloaded.
  * @returns The fluidBuild section of the package.json.
  */
+export function loadFluidBuildConfig(rootDir: string, noCache = false): IFluidBuildConfig {
```
Is the only difference between the load and get functions that load throws and get doesn't? Can they be consolidated and calling code updated to work consistently (i.e. always check undefined or handle exception)?

curtisman replied:
I could, but didn't want to change too many code site.

curtisman replied:
I was thinking of rewriting to whole init/infer flow to clean it up with some fluid repo specific behavior, and reduce multiple calls to `getPackages` and ways to detect package managers, but didn't want to churn the code with this PR and impact other tools that share this code.   We can talk about how (and when) we want to clean this up.

tylerbutler commented on build-tools/packages/build-tools/src/fluidBuild/fluidRepoBuild.ts:
```diff
@@ -32,8 +32,17 @@ export interface IPackageMatchedOptions {
 export class FluidRepoBuild extends FluidRepo {
-	constructor(resolvedRoot: string) {
-		super(resolvedRoot);
+	public static create(resolvedRoot: string) {
```
What's the advantage of static create vs. constructor?

curtisman replied:
Just stylistic.  I am used to have "super" call be always the first thing in a constructor.

Also, it will mirror FluidRepo creation pattern.

tylerbutler reviewed (commented):
Mostly confused by the get/load changes, but I also am just returning from being out, so I'm rusty. :)

curtisman committed:
message

tylerbutler committed:
Merge branch 'main' into rootr

curtisman committed:
fluid-build: support more `copyfiles` flags and behavior (#19949)

- Support -a and -F flags
- Support multiple source.

Also, warn if a glob pattern isn't quoted, which has different behavior
in different shell and OS

tylerbutler committed:
subpodule revert
